### DISCUSSION TITLE
[@lit-labs/context] Context not available when children are not part of render() method

### DISCUSSION THREAD

**Discussion by BregtDeLange** [lit/lit#3790] [Q&A]
Just started getting into more complex usages for Lit, and found out about the @lit-labs/context

I have an API that returns product data that I want to make available to different components. I used to to this with events, and children subscribing to the event that is created at root level. However, my components are without Shadow DOM, because I want to style them from the Light DOM.

I can get my data working with the Context setup, when I have my child web components are part of the render method. But that's not how I use my web components in this project. The root element does not define how the HTML structure of its children should be. I want my children to be dependent on the root in terms of data and functionality, but I don't want the HTML to be forced from the render() method on the root. I want my child elements be placed inside elements chosen in the light DOM.

Here's an example

```
// root.ts
@customElement("my-root")
export class RootComponent extends LitElement {
    @provide({ context: myContext})
    data: any = {};

    connectedCallback(): void {
        super.connectedCallback();

        const api = new Api({...});

        api.getData().then((data) => {
            this.data = data;
        });
    }

    render() {
        return html`<slot></slot>`;
    }
}

// **child-component.ts**
@customElement("my-child")
class ChildComponent extends LitElement {
    @consume({ context: myContext, subscribe: true })
    data: any;

    // Turn off Shadow DOM
    createRenderRoot() {
        return this;
    }

    render() {
        return html`${this.data?.something?.map((item) => html`<other-component .item=${item}></other-component>`)}`;
    }
}

@customElement("other-child")
class OtherComponent extends LitElement {
    @consume({ context: myContext, subscribe: true })
    data: any;

    // Turn off Shadow DOM
    createRenderRoot() {
        return this;
    }

    render() {
        return html`${this.data?.somethingElse?.map((item) => html`<fancy-component .item=${item}></fancy-component>`)}`;
    }
}

// index.html: my data context is empty
<my-root>
    <section>
        <!-- I want to choose where my-child component is positioned on my page (only requirement is to be part of app-root -->
        <my-child class="something"><my-child>
    </section>
</my-root>

// other-page.html: I need a different layout, and want to add the other-child component, but my data context is empty
<my-root>
    <div class="my-grid">
        <other-child class="something"><other-child>
        <my-child class="something-else"><my-child>
    </div>
</my-root>
```

However, it obviously works when I do this in my root.ts, but that is not really how my site works right now. I want more freedom with the child components.

```
// root.ts
...
    render() {
        return html`<my-child></my-child><other-child></other-child>`;
    }
```

Did I forgot something in my desired setup? Or is that not possible with the @lib-labs/context package? I'd assumed that all children could subscribe to the events set by the context object, even if they're not part of the actual render method.

**Comment by augustjk** [comment]
Slotted children components of a context provider can still consume the context. See https://lit.dev/playground/#gist=6b69a35321ee746605f877a5ed03ee8d

If it's not working for you, we'd need to see a reproduction.
